File: v1/LLD/Design-Patterns/TemplateMethod/code/template_method.js
/**
 * Template Method — CaffeineBeverage fixes the recipe sequence; Tea and Coffee fill in the varying steps.
 * Run: node template_method.js
 */

class CaffeineBeverage {
  // The template method — the fixed sequence of steps.
  prepareRecipe() {
    this.boilWater();
    this.brew();
    this.pourInCup();
    if (this.wantsCondiments()) {
      this.addCondiments();
    }
  }

  boilWater() {
    console.log("Boiling water");
  }

  pourInCup() {
    console.log("Pouring into cup");
  }

  brew() {
    throw new Error("brew() must be implemented by subclass");
  }

  addCondiments() {
    throw new Error("addCondiments() must be implemented by subclass");
  }

  // Hook — subclasses may override; default is true.
  wantsCondiments() {
    return true;
  }
}

class Tea extends CaffeineBeverage {
  brew() {
    console.log("Steeping the tea");
  }

  addCondiments() {
    console.log("Adding lemon");
  }
}

class Coffee extends CaffeineBeverage {
  brew() {
    console.log("Dripping coffee through filter");
  }

  addCondiments() {
    console.log("Adding sugar and milk");
  }

  wantsCondiments() {
    return false;
  }
}

console.log("Preparing tea:");
const tea = new Tea();
tea.prepareRecipe();

console.log("Preparing coffee (no condiments):");
const coffee = new Coffee();
coffee.prepareRecipe();

module.exports = { CaffeineBeverage, Tea, Coffee };

File: v1/LLD/Design-Patterns/TemplateMethod/code/TemplateMethod.java
// Template Method — CaffeineBeverage fixes the recipe sequence; Tea and Coffee fill in the varying steps.
// Compile: javac TemplateMethod.java
// Run:     java TemplateMethod

abstract class CaffeineBeverage {
    // The template method — final so subclasses can't change the sequence.
    final void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (wantsCondiments()) {
            addCondiments();
        }
    }

    void boilWater() {
        System.out.println("Boiling water");
    }

    void pourInCup() {
        System.out.println("Pouring into cup");
    }

    abstract void brew();
    abstract void addCondiments();

    // Hook — subclasses may override; default is "yes".
    boolean wantsCondiments() {
        return true;
    }
}

class Tea extends CaffeineBeverage {
    void brew() {
        System.out.println("Steeping the tea");
    }

    void addCondiments() {
        System.out.println("Adding lemon");
    }
}

class Coffee extends CaffeineBeverage {
    void brew() {
        System.out.println("Dripping coffee through filter");
    }

    void addCondiments() {
        System.out.println("Adding sugar and milk");
    }

    // Overrides the hook — this coffee is prepared black.
    boolean wantsCondiments() {
        return false;
    }
}

public class TemplateMethod {
    public static void main(String[] args) {
        System.out.println("Preparing tea:");
        CaffeineBeverage tea = new Tea();
        tea.prepareRecipe();

        System.out.println("Preparing coffee (no condiments):");
        CaffeineBeverage coffee = new Coffee();
        coffee.prepareRecipe();
    }
}
